Stub usage report implementation

jaydeepk opened the pull request:
**What**:

Implemented the generation of a 'Stub Usage Report' json file.
The report will be automatically generated once the stub process is stopped.
The report will list all the endpoints for which the stub returned a successful response (in a format similar to the API Coverage Report).


- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [x] Tests
- [ ] Sonar Quality Gate

jaydeepk committed:
Recorded all successful stub requests and responses ( to be further used for generating the Stub Usage Report)

jaydeepk committed:
Added source properties to StubRequestLog.
Added StubUsageReport class to generate the report.

jaydeepk committed:
Implemented the generation of the Stub Usage Report based on a list all the apis from specifications and the recorded request logs.

jaydeepk committed:
Added Feature scenarios to Stub API list.

jaydeepk committed:
Implemented generation of Stub Usage report when Specmatic is invoked via command line.
If stub contracts are explicitly passed as arguments, the stub usage report will not be generated.

jaydeepk committed:
Refactored extracting source repository properties from HttpStubResponse

jaydeepk committed:
Minor refactoring, renamed transient and non transient stub queues

jaydeepk committed:
Added matched Feature and Scenario to HttpStubData so that they can be used to extract source repository info while recording logs

jaydeepk committed:
Added getters for exposing logs and stubApis in HttpStub.
Added tests to validate stub logs in different scenarios like stubbed response, fake response, error response

jaydeepk committed:
Stub usage printing moved to the close method.
Removed the printUsageReport method from ContractStub

jaydeepk committed:
Minor refactoring/renaming

jaydeepk committed:
Added tests to validate stub usage report json file, strict mode.
Minor refactoring.

jaydeepk committed:
minor change

harikrishnan83 commented on core/src/main/kotlin/in/specmatic/core/HttpResponse.kt:
```diff
@@ -35,6 +35,8 @@ data class HttpResponse(
     fun specmaticResultHeaderValue(): String =
         this.headers.getOrDefault(SPECMATIC_RESULT_HEADER, "success")
 
+    fun isSuccessful(): Boolean = specmaticResultHeaderValue() == "success"
```
@joelrosario is there another way to check if a response is a success response? also just curious if the result header value is a dependable way of checking if a response is a success response. cc @jaydeepk

joelrosario replied:
I think there's a way we can use the type system to signal success or failure.

The code is structured to return a stubbed response. However it's possible to wrap the HttpStubResponse object in another object that contains the behaviour for logging (or not logging as the case may be).

The most expeditious way would be to create a wrapper function around HttpStub/getHttpResponse (or alter it's signature), then introduce the wrapper object in getHttpResponse at line 624, and in fakeHttpResponse at lines 756 and 758.

cc @jaydeepk

harikrishnan83 replied:
@jaydeepk @joelrosario okay, if this is dependable way of checking if the response is a success then looks good.

joelrosario commented on core/src/main/kotlin/in/specmatic/stub/HttpStub.kt:
```diff
@@ -427,6 +459,40 @@ class HttpStub(
 
     init {
         server.start()
+        extractALlEndpoints()
```
I think this method can be called when _allEndpoints is initialised, and then _allEndpoints can become immutable

joelrosario committed:
Made endpoint extraction logic in HttpStub immutable

joelrosario committed:
Removed dead code

joelrosario commented on core/src/main/kotlin/in/specmatic/core/HttpResponse.kt:
```diff
@@ -35,6 +35,8 @@ data class HttpResponse(
     fun specmaticResultHeaderValue(): String =
         this.headers.getOrDefault(SPECMATIC_RESULT_HEADER, "success")
```
Should we not default to "failure"? If the header does not exist it means failure.

joelrosario replied:
This is also used elsewhere in the code in the context of which this behaviour is correct. Closing this.

joelrosario committed:
Log the path of ignored file (instead of object)

joelrosario reviewed (changes requested):
I've left some comments.

joelrosario commented on core/src/main/kotlin/in/specmatic/stub/HttpStub.kt:
```diff
@@ -65,9 +67,21 @@ class HttpStub(
     private val threadSafeHttpStubs = ThreadSafeListOfStubs(_httpStubs.filter { it.stubToken == null }.toMutableList())
     private val threadSafeHttpStubQueue = ThreadSafeListOfStubs(_httpStubs.filter { it.stubToken != null }.reversed().toMutableList())
 
+    private val _logs: MutableList<StubEndpoint> = mutableListOf()
```
This data structure needs to be made thread-safe

jaydeepk replied:
Done

joelrosario committed:
Use the type system to signal whether a stub was found

We need to check if a stub match was found so that we know whether to
track it for coverage. The code is currently structured to return
responses. But with this change we are returning a stub result object
indicating success or failure, and we've moved the logging behaviour
into that object.

joelrosario committed:
Extracted stubbed response classes to separate files

joelrosario committed:
Moved the stub coverage tests into a new file

jaydeepk committed:
Made the logs collection threadsafe.

joelrosario commented on core/src/main/kotlin/in/specmatic/stub/HttpStub.kt:
```diff
@@ -65,9 +67,21 @@ class HttpStub(
     private val threadSafeHttpStubs = ThreadSafeListOfStubs(_httpStubs.filter { it.stubToken == null }.toMutableList())
     private val threadSafeHttpStubQueue = ThreadSafeListOfStubs(_httpStubs.filter { it.stubToken != null }.reversed().toMutableList())
 
+    private val _logs: MutableList<StubEndpoint> = Collections.synchronizedList(ArrayList())
```
There should be a test for this. Have say 10 threads running simultaneously making requests, and we should have 10 logs in this list.

jaydeepk replied:
Sure, I have added the test.

jaydeepk committed:
Added test to ensure that concurrent request are logged successfully by the stub.

joelrosario reviewed (approved):
Looks good to me, thanks @jaydeepk
